OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,274 @@
+//=============================================================================
+
+
+/** Concrete Implementation of the <tt>DataTable<DataType></tt>
+@tparam DataType  Any numerical type: (SimTK::) Real, Vec<M>, Mat<M,N>, ...*/
+template<typename DataType = SimTK::Real>
+class DataTable_ : public AbstractDataTable {
+//=============================================================================
+// METHODS
+//=============================================================================
+public:
+    virtual ~DataTable_(){};
+
+    /** Construct an empty DataTable */
+    DataTable_() : _type(typeid(DataType)) {}
+
+    /** Construct a pre-sized DataTable to be filled in later. If values
+        are accessed before they are assigned they will be NaN. */
+    DataTable_(size_t nrows, size_t ncols, 
+        const DataType& val = DataType(SimTK::NaN)) :
+        _data(int(nrows), int(ncols), val),
+        _type(typeid(DataType)) {}
+
+    /** Convenience construction of a DataTable from a filename with its source
+        type identified by the file extension. E.g. `results.csv`, `states.sto`
+    @param[in] filename         the name of the data file to be accessed
+    */
+    DataTable_(const std::string& filename) :
+        _type(typeid(DataType)) {
+        std::string daId = FileAdapter::findExtension(filename);
+        auto* reader = FileAdapter::createAdapter(daId);
+        if (!reader){
+            throw Exception("DataTable() no adapter to read file '"
+                + filename + "' could be found.");
+        }
+        reader->setFilename(filename);
+        reader->prepareForReading(*this);
+        reader->read();
+        delete reader;
+    }
+
+    DataTable_& operator=(const DataTable_& dt) {
+        _metaData = dt.getMetaData();
+        _data = dt.getAsMatrix() ;
+        _columnLabels = dt.getColumnLabels();
+        return *this;
+    }
+
+    virtual AbstractDataTable* clone() const override {
+        return new DataTable_(*this);
+    }
+
+    const XMLDocument& getMetaData() const override final { return _metaData; }
+    XMLDocument& updMetaData() override final { return _metaData; }
+
+    size_t  getNumRows() const override { return _data.nrow(); }
+    size_t  getNumCols() const override { return _data.ncol(); }
+
+    const Array<std::string>& getColumnLabels() const override {
+        return _columnLabels;
+    }
+
+    Array<std::string>& updColumnLabels() override {
+        return _columnLabels;
+    }
+    
+    /** Return true if the DataTable contains a column for the label. */
+    bool hasColumn(const std::string& label) const override{
+        return (_columnLabels.findIndex(label) >= 0);
+    }
+
+    size_t getColumnIndex(const std::string& label) const override {
+        return _columnLabels.findIndex(label);
+    }
+
+    /** Read-only access to a row of data in the table as a RowVector.*/
+    const SimTK::RowVectorView_<DataType> getRow(size_t row) const {
```

Comment:
> It is probably meant to be a const T& and not const T. I think it would be inefficient to copy.

This one is OK; the "View" in RowVectorView means it is just a small handle containing a pointer to the real data, with indexing (stride) information. To return a const& would require us to store RowVectorView structures in the DataTable.